auditable-serde/src/lib.rs

```diff
@@ -133,25 +136,44 @@ pub struct Package {
     pub root: bool,
 }
 
-/// Serializes to "git", "local", "crates.io" or "registry". Designed to be extensible with other revision control systems, etc.
+/// Serializes to "git", "local", "crates.io", "registry" or a more complex
+/// struct with any of those values in the `kind` field. Designed to be
+/// extensible with other revision control systems, etc.
+//
+// The abundance of schemars attributes was introduced to fix an unexpected
+// way of representing untagged enums that is inconsistent with serde. Without
+// extra `with` attributes the generated schema assigns null types to instances
+// of the enum's variants which are unit types instead of using string type.
 #[non_exhaustive]
 #[derive(Serialize, Deserialize, Debug, PartialEq, Eq, PartialOrd, Ord, Clone)]
-#[serde(from = "&str")]
-#[serde(into = "String")]
+#[serde(rename_all = "snake_case", untagged)]
 #[cfg_attr(feature = "schema", derive(schemars::JsonSchema))]
 pub enum Source {
+    /// "crates.io"
+    #[serde(rename = "crates.io")]
+    #[cfg_attr(feature = "schema", schemars(with = "String"))]
     CratesIo,
-    Git,
+    /// "local"
+    #[cfg_attr(feature = "schema", schemars(with = "String"))]
     Local,
+    /// "registry"
+    #[cfg_attr(feature = "schema", schemars(with = "String"))]
     Registry,
+    #[serde(with = "compact_enum_variant")]
+    #[cfg_attr(
+        feature = "schema",
+        schemars(schema_with = "compact_enum_variant::schema::<Source, GitSource>",)
+    )]
+    Git(GitSource),
```

Comment:
This specific fragment changes the `Git` variant to use the representation specified in the `compact_enum_variant.rs` file. The attributes did became a little bit complex because of that but besides most of the complexity should be hidden away in that file and in future a similar trick can be used to extend any other enum variant while keeping it all backwards compatible with the way the unit variants are serialized.

That is, a change from:
```rs
#[derive(Serialize, Deserialize)]
enum MyEnum{
 V1,
}
```
to:
```rs
#[derive(Serialize, Deserialize)]
#[serde(untagged)]
enum MyEnum {
 #[serde(with = "compact_enum_variant")]
 V1(V1Struct),
}

/// And some impls ... {}
```

Should be backwards compatible.